### PR TITLE
Jordan/pow cache management

### PR DESCRIPTION
**Pull request by mjfh**

### CODE REVIEW

File: nimbus/utils/pow.nim [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+  if full.isSome:
+    tm.fullByEpoch = full.get
+  else:
+    tm.fullByEpoch = PowDatasetRef.init(cache = tm.lightByEpoch)
+
+# ------------------------------------------------------------------------------
+# Public functions, Constructor
+# ------------------------------------------------------------------------------
+
+proc init*(T: type PowRef;
```
Comment (jangko) [inline nimbus/utils/pow.nim:187]: It's a bit unusual for a ref object constructed using `init`. `init` usually used  for regular object, or to init ref object like the one in line 163. overloading `new` will be more consistent.

File: nimbus/utils/pow.nim [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+            rng: ref BrHmacDrbgContext;
+                   cache: PowCacheRef;
+                   dataset: PowDatasetRef): T =
+  ## Constructor
+  new result
+  result.init(some(rng), some(cache), some(dataset))
+
+proc init*(T: type PowRef; cache: PowCacheRef; dataset: PowDatasetRef): T =
```
Comment (jangko) [inline nimbus/utils/pow.nim:195]: ditto

File: nimbus/utils/pow.nim [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  ## Constructor
+  new result
+  result.init(none(ref BrHmacDrbgContext), some(cache), some(dataset))
+
+proc init*(T: type PowRef; rng: ref BrHmacDrbgContext): T =
```
Comment (jangko) [inline nimbus/utils/pow.nim:200]: ditto

File: nimbus/utils/pow.nim [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+  ## Constructor
+  new result
+  result.init(some(rng), none(PowCacheRef), none(PowDatasetRef))
+
+proc init*(T: type PowRef): T =
```
Comment (jangko) [inline nimbus/utils/pow.nim:205]: ditto

File: nimbus/utils/pow/pow_cache.nim [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+proc init*(pc: var PowCache; maxItems = nItemsMax) =
+  ## Constructor for PoW cache
+  pc.cacheMax = maxItems
+  pc.cache.init(nItemsInit)
+
+proc init*(T: type PowCache; maxItems = nItemsMax): T =
+  ## Constructor variant
+  result.init(maxItems)
+
+proc init*(T: type PowCacheRef; maxItems = nItemsMax): T =
```
Comment (jangko) [inline nimbus/utils/pow/pow_cache.nim:64]: ditto

File: nimbus/utils/pow/pow_dataset.nim [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+proc init*(T: type PowDataset; maxItems = nItemsMax; cache: PowCacheRef): T =
+  ## Constructor variant
+  result.init(some(maxItems), some(cache))
+
+proc init*(T: type PowDataset; maxItems = nItemsMax): T =
+  ## Constructor variant
+  result.init(some(maxItems), none(PowCacheRef))
+
+
+proc init*(T: type PowDatasetRef; maxItems = nItemsMax; cache: PowCacheRef): T =
```
Comment (jangko) [inline nimbus/utils/pow/pow_dataset.nim:91]: ditto

File: nimbus/utils/pow/pow_dataset.nim [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  ## Constructor variant
+  new result
+  result[].init(some(maxItems), some(cache))
+
+proc init*(T: type PowDatasetRef; maxItems = nItemsMax): T =
```
Comment (jangko) [inline nimbus/utils/pow/pow_dataset.nim:96]: ditto

**Review by jangko — CHANGES_REQUESTED** [review]

File: nimbus/utils/pow/pow_dataset.nim (hunk 7 above)
Comment (mjfh) [reply nimbus/utils/pow/pow_dataset.nim]: ok

**Comment by jangko** [comment]
need rebase because of nim-json-rpc updates

**Comment by mjfh** [comment]
ok thanks :)

**Pull request merged**